tests/Collection/CollectionFunctionalTest.php

```diff
@@ -739,7 +748,16 @@ public function testMethodDoesNotInheritReadWriteConcernInTranasaction(Closure $
 
         (new CommandObserver())->observe(
             function () use ($method, $collection, $session): void {
-                call_user_func($method, $collection, $session);
+                try {
+                    call_user_func($method, $collection, $session);
+                } catch (CommandException $exception) {
+                    // Write aggregates are not supported in transactions
+                    if ($exception->getResultDocument()->codeName === 'OperationNotSupportedInTransaction') {
+                        $this->markTestSkipped('OperationNotSupportedInTransaction: ' . $exception->getMessage());
```

Comment:
`$out` and `$merge` operators aren't supported in transaction. They throw a server error. I have a doubt on the best solution to exclude the new "read-write aggregate" method:
1. This solution: skip the test in case of exception. But unlike tests skipped due to a temporary incompatiblity with the running environment, this test will never pass.
2. Ignores the error and still run the assertions on command options. Even if that does not make sense to test a failing command. That could hide other errors and make debugging more difficult.
3. exclude this case from the test cases by adding a `collectionTransactionMethodClosures` data provider and add a new char `t` in `$rw` to exclude methods incompatible with transactions.

Reply:
> 1. This solution: skip the test in case of exception. But unlike tests skipped due to a temporary incompatiblity with the running environment, this test will never pass.
>
> 2. Ignores the error and still run the assertions on command options. Even if that does not make sense to test a failing command. That could hide other errors and make debugging more difficult.

I don't think conditionally skipping the test makes sense, given that the test only cares about asserting that `readConcern` and `writeConcern` are omitted on the outgoing command.

Catching and ignoring [`OperationNotSupportedInTransaction`](https://github.com/mongodb/mongo/blob/master/src/mongo/base/error_codes.yml#LL312C57-L312C57) seems preferable. For consistency, I would suggest defining a private constant for its numeric code (263).

That said, I don't think that's a permanent solution. Read on.

> 3. exclude this case from the test cases by adding a collectionTransactionMethodClosures data provider and add a new char t in $rw to exclude methods incompatible with transactions.

I don't think this is necessary. `collectionMethodClosures()` already comments out various methods that cannot be run within a transaction. I was going to suggest removing all of those comments after adding logic to ignore OperationNotSupportedInTransaction, but then I realized that would only fix `testMethodDoesNotInheritReadWriteConcernInTransaction`.

`testMethodInTransactionWithWriteConcernOption` and `testMethodInTransactionWithReadConcernOption` use derivative data providers that filter for strings containing "w" and "r", respectively. Those tests would never work with the unsupported operations because we never added logic to check for active transactions and throw UnsupportedException.

Aggregate is a special case because pipelines are generally supported in transactions and only write operators are unsupported.

----

With all of that in mind, I think the most sensible solution is removing the pipeline with a write stage and just leaving a comment above read-only pipeline to explain that we're intentionally not testing a pipeline with a write stage since those are unsupported.

That would mean we have no test coverage for the following logic near the top of `Aggregate::execute()`:

```php
if (isset($this->options['writeConcern'])) {
   throw UnsupportedException::writeConcernNotSupportedInTransaction();
}
```

If you'd like to cover that, I'd suggest doing one of the following:

 * Create a second data provider for `testMethodDoesNotInheritReadWriteConcernInTransaction()` that only includes the write pipeline. The test only takes one parameter so there'd be no reason to include "rw" in the data provider. This would still require modifying the test to catch and ignore OperationNotSupportedInTransaction errors.
 * Create a separate test similar to `testMethodDoesNotInheritReadWriteConcernInTransaction()` that tests Aggregate directly with the write pipeline. A doc block for that test would be a good opportunity to explain why we couldn't do this with the common data provider.

Alternatively, just remove the write pipeline and leave this untested. I'm not too concerned about doing so since we know this code path will lead to an exception anyway.

Reply:
You're right, there's no point in testing behavior on a command that's going to fail anyway.